File: core/src/main/java/com/jinkyumpark/bookitout/search/apiResponse/seoulLibrary/ApiSeoulLibraryBook.java
package com.jinkyumpark.bookitout.search.apiResponse.seoulLibrary;

import com.jinkyumpark.bookitout.search.provider.OnlineLibraryProvider;
import com.jinkyumpark.bookitout.search.response.searchResult.OnlineLibrarySearchResult;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;

@NoArgsConstructor @AllArgsConstructor
@Getter
public class ApiSeoulLibraryBook {
    private String contentsKey;
    private String title;
    private String author;
    private String publisher;
    private String publishDate;
    private String contentsType;
    private String contentsTypeDesc;
    private String loanCnt;
    private String ownerCode;
    private String coverUrl;
    private String coverMSizeUrl;
    private String coverSSizeUrl;
    private String contentsInfo;
    private String recommendCnt;
    private String reserveCnt;
    private String reviewRate;
    private String likeYn;
    private String isbn;
    private Integer currentLoanCount;
    private Integer currentResvCount;
    private Integer reviewCount;
    private Integer likeCount;
    private Double avgrate;

    public OnlineLibrarySearchResult toOnlineLibrarySearchResult() {
        return OnlineLibrarySearchResult.builder()
                .title(title)
                .author(author)
                .cover(coverUrl)
                .link(String.format("https://elib.seoul.go.kr/contents/detail.do?no=%s", contentsKey))
                .loanPossible(currentLoanCount < 5)
                .reservationPossible(currentResvCount < 5)
                .provider(OnlineLibraryProvider.SEOUL_LIBRARY)
                .build();
    }
}

File: search/src/main/kotlin/com/jinkyumpark/search/apiResponse/seoulLibrary/ApiSeoulLibraryBook.kt
package com.jinkyumpark.search.apiResponse.seoulLibrary

import com.jinkyumpark.search.common.SearchProvider
import com.jinkyumpark.search.common.SearchResult

data class ApiSeoulLibraryBook(
    val contentsKey: String,
    val title: String,
    val author: String,
    val publisher: String,
    val publishDate: String,
    val contentsType: String,
    val contentsTypeDesc: String,
    val loanCnt: String,
    val ownerCode: String,
    val coverUrl: String,
    val coverMSizeUrl: String,
    val coverSSizeUrl: String,
    val contentsInfo: String,
    val recommendCnt: String,
    val reserveCnt: String,
    val reviewRate: String,
    val likeYn: String,
    val isbn: String,
    val currentLoanCount: Int,
    val currentResvCount: Int,
    val reviewCount: Int,
    val likeCount: Int,
    val avgrate: Double,
) {

    fun toOnlineLibraryResponse(): SearchResult {
        return SearchResult(
            title = title,
            author = author,
            cover = coverUrl,
            link = "https://elib.seoul.go.kr/contents/detail.do?no=$contentsKey",
            loanPossible = currentLoanCount < 5,
            reservationPossible = currentResvCount < 5,
            searchProvider = SearchProvider.SEOUL_LIBRARY,
        )
    }

}
